File: cpp/src/memory-leak/memory_leak_safe.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

//----------------------------------Szenario------------------------------------------
/*
 * In diesem Programm wird ein Memory Leak durch die Verwendung von `std::weak_ptr`
 * verhindert.
 * Die Tasks speichern nur schwache Referenzen (`std::weak_ptr`) auf den Scheduler,
 * wodurch der Referenzzyklus zwischen `Scheduler` und `Task` aufgelöst wird.
 * Der Scheduler nutzt `std::enable_shared_from_this`, um eine starke Referenz
 * zu sich selbst zu erstellen, die sicher an die Tasks übergeben wird.
 */

class Scheduler; // Vorwärtsdeklaration

class Task
{
public:
    explicit Task(const std::string& name) : name(name)
    {
        std::cout << "Task \"" << name << "\" erstellt.\n";
    }

    ~Task() { std::cout << "Task \"" << name << "\" gelöscht.\n"; }

    // Setzt den Scheduler für diese Task
    void setScheduler(const std::shared_ptr<Scheduler>& sched)
    {
        scheduler = sched; // `weak_ptr` speichert keine starke Referenz
    }

    // Gibt den Namen der Task aus
    void printName() const { std::cout << "Task Name: " << name << '\n'; }

private:
    std::string name;
    std::weak_ptr<Scheduler> scheduler; // Schwache Referenz auf Scheduler
};

class Scheduler : public std::enable_shared_from_this<Scheduler>
{
public:
    Scheduler() { std::cout << "Scheduler erstellt.\n"; }

    ~Scheduler() { std::cout << "Scheduler gelöscht.\n"; }

    // Fügt eine Task zum Scheduler hinzu und setzt den Scheduler für die Task
    void addTask(const std::shared_ptr<Task>& task)
    {
        tasks.push_back(task);
        task->setScheduler(
            shared_from_this()); // Stellt sicher, dass eine schwache Referenz gesetzt wird
    }

private:
    std::vector<std::shared_ptr<Task>> tasks; // Liste von Tasks
};

int main()
{
    // Erstelle einen Scheduler und Tasks
    auto scheduler = std::make_shared<Scheduler>();
    auto task1 = std::make_shared<Task>("Task 1");
    auto task2 = std::make_shared<Task>("Task 2");

    // Füge die Tasks dem Scheduler hinzu
    scheduler->addTask(task1);
    scheduler->addTask(task2);

    // Gebe die Namen der Tasks aus
    task1->printName();
    task2->printName();

    // Debug-Ausgabe für Referenzzähler
    std::cout << "Referenzzähler des Schedulers: " << scheduler.use_count() << '\n';

    return 0; // Kein Memory Leak, da schwache Referenzen genutzt werden
}

File: rust/memory-leak/src/memory_leak_safe.rs
use std::cell::RefCell;
use std::rc::{Rc, Weak};

//----------------------------------Szenario------------------------------------------
/*
 * In diesem Programm wird ein Memory Leak durch die Verwendung von `Rc` und `Weak`
 * verhindert.
 * Der `Task` hält eine schwache Referenz (`Weak<Scheduler>`) auf den `Scheduler`,
 * wodurch der Referenzzyklus zwischen `Scheduler` und `Task` aufgelöst wird.
 * Der `Scheduler` verwaltet die Tasks mit starken Referenzen (`Rc<Task>`), während die
 * Rückverweise der Tasks auf den Scheduler den Referenzzähler nicht erhöhen.
 */

// Repräsentiert eine einzelne Aufgabe (`Task`) innerhalb eines Schedulers.
// Jede Aufgabe kennt ihren Namen und besitzt eine schwache Referenz (`Weak<Scheduler>`)
// auf den Scheduler, um zyklische Referenzen zu vermeiden.
struct Task {
    name: String,
    scheduler: RefCell<Option<Weak<Scheduler>>>, // Schwache Referenz auf den Scheduler
}

impl Task {
    // Erstellt eine neue `Task` mit dem gegebenen Namen und gibt eine `Rc<Task>`-Referenz zurück.
    fn new(name: &str) -> Rc<Task> {
        Rc::new(Task {
            name: name.to_string(),
            scheduler: RefCell::new(None), // Anfangs keine Verbindung zum Scheduler
        })
    }
    // Setzt den Scheduler für die Aufgabe, indem `Rc::downgrade` verwendet wird,
    // um eine `Weak<Scheduler>`-Referenz zu erstellen.
    fn set_scheduler(&self, scheduler: Rc<Scheduler>) {
        *self.scheduler.borrow_mut() = Some(Rc::downgrade(&scheduler));
    }

    /// Gibt den Namen der Task aus.
    fn print_name(&self) {
        println!("Task Name: {}", self.name);
    }
}
// Der `Scheduler` verwaltet eine Liste von `Task`-Instanzen.
// Die `tasks`-Liste speichert `Rc<Task>`-Referenzen, wodurch sichergestellt wird,
// dass die Aufgaben solange existieren, wie sie vom Scheduler verwaltet werden.
struct Scheduler {
    tasks: RefCell<Vec<Rc<Task>>>, // Initial leerer Task-Container
}

impl Scheduler {
    fn new() -> Rc<Scheduler> {
        Rc::new(Scheduler {
            tasks: RefCell::new(Vec::new()),
        })
    }
    // Fügt eine `Task` dem Scheduler hinzu.
    fn add_task(&self, task: Rc<Task>) {
        self.tasks.borrow_mut().push(task);
    }

    // Gibt alle verwalteten Tasks aus.
    fn print_tasks(&self) {
        println!("Scheduler verwaltet folgende Tasks:");
        for task in self.tasks.borrow().iter() {
            task.print_name();
        }
    }
}

fn main() {
    // Erstellt eine Instanz des `Scheduler`
    let scheduler = Scheduler::new();

    // Erstellt zwei Tasks
    let task1 = Task::new("Task 1");
    let task2 = Task::new("Task 2");

    // Fügt die Tasks zum Scheduler hinzu
    scheduler.add_task(task1.clone());
    scheduler.add_task(task2.clone());

    // Setzt den Scheduler für die Tasks, wobei eine schwache Referenz verwendet wird
    task1.set_scheduler(scheduler.clone());
    task2.set_scheduler(scheduler.clone());

    // Gibt alle Aufgaben aus, die der Scheduler verwaltet
    scheduler.print_tasks();

    // Debugging: Gibt die Referenzzähler für den Scheduler und die Tasks aus
    println!("Scheduler und Tasks erstellt.");
    println!(
        "Referenzzähler des Schedulers: {}",
        Rc::strong_count(&scheduler)
    );
    println!("Referenzzähler von Task 1: {}", Rc::strong_count(&task1));
}
